### PR TITLE
.Net 6 auth with specific scheme /3

### LINKED ISSUE #16389: Multiple authentication schemes
I am trying to create a sample application using multiple authentication schemes (AAD B2C and another AAD tenant). If I configure only one of them (does not matter which one) it works like a charm and everything works correctly. If I configure both of them (as in the provided example), however mention only one of them in the default policy, it also works (only for that scheme):
```
services.AddAuthentication()
.AddJwtBearer("B2C", jwtOptions =>
{
    jwtOptions.Authority = "my-b2c-authority";
    jwtOptions.Audience = "my-audience";
})
.AddJwtBearer("AAD", jwtOptions =>
{
    jwtOptions.Authority = "my-aad-authority";
    jwtOptions.Audience = "my-audience";
});
```

```
services.AddAuthorization(options =>
{
    options.DefaultPolicy = new AuthorizationPolicyBuilder()
        .RequireAuthenticatedUser()
        .AddAuthenticationSchemes("B2C")
        .Build();
});
```
However, if I configure both and mention them in the default policy:
```
services.AddAuthorization(options =>
{
    options.DefaultPolicy = new AuthorizationPolicyBuilder()
        .RequireAuthenticatedUser()
        .AddAuthenticationSchemes("B2C", "AAD")
        .Build();
});
```
it seems that the token gets validated twice (against each configured scheme) and therefore it fails - I am suspecting that as I can see the following messages in the debug window:
```
Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler:Information: Successfully validated the token.
...
Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler:Information: Failed to validate the token.

Microsoft.IdentityModel.Tokens.SecurityTokenSignatureKeyNotFoundException: IDX10501: Signature validation failed. Unable to match keys: 
kid: '[PII is hidden]', 
token: '[PII is hidden]'.
   at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateSignature(String token, TokenValidationParameters validationParameters)
   at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateToken(String token, TokenValidationParameters validationParameters, SecurityToken& validatedToken)
   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
...
```

I guess I am missing something and I have not been able to figure it out on my own. Would be grateful for any hints regarding what might be wrong.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ee71441f-471a-89cd-0ef4-9b32b5dd85cf
* Version Independent ID: 0096c365-2f4e-5d21-89a0-c4f10fcb0ed9
* Content: [Authorize with a specific scheme in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-3.1)
* Content Source: [aspnetcore/security/authorization/limitingidentitybyscheme.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/authorization/limitingidentitybyscheme.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### LINKED ISSUE #24605: .NET 6: Authorize with a specific scheme
Also do #16389

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ee71441f-471a-89cd-0ef4-9b32b5dd85cf
* Version Independent ID: 0096c365-2f4e-5d21-89a0-c4f10fcb0ed9
* Content: [Authorize with a specific scheme in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-6.0)
* Content Source: [aspnetcore/security/authorization/limitingidentitybyscheme.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/authorization/limitingidentitybyscheme.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #24605 Fixes #16389
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-6.0&branch=pr-en-us-24623#use-multiple-authentication-schemes)

## Public review  of the doc build 
- download the [HTML files](https://www.dropbox.com/sh/yy994551cje0kap/AAB0tRnm7gze2kYrSYPSFHIda?dl=0)
- unzip
- view HTML file in a browser. It best to view the doc with the widest window that has no left or right TOC (Table  of contents)
- For links outside the doc, remove the `review.` from URL.

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@dandax @egbakou @Sayan751 @jamir-araujo @automachina please review, especially 
## Use multiple authentication schemes

**Comment by jamir-araujo** [comment]
@Rick-Anderson, I notice that you mention an issue with the `IDX10501` problem, but this issue doesn't have the solutions posted on the [#16389](https://github.com/dotnet/AspNetCore.Docs/issues/16389) issue.

**Comment by Rick-Anderson** [comment]
> @Rick-Anderson, I notice that you mention an issue with the `IDX10501` problem, but this issue doesn't have the solutions posted on the [#16389](https://github.com/dotnet/AspNetCore.Docs/issues/16389) issue.

It does according to @dandax [in his comment](https://github.com/dotnet/AspNetCore.Docs/issues/16389#issuecomment-984964222)

**Comment by jamir-araujo** [comment]
@Rick-Anderson, with @dandax code the application will work, but you will most likely have the `IDX10501` in the application logs. Being that the two schemes have different authorities, a jwt issued by one will not be valid in the other. At least that's my experience.

**Comment by egbakou** [comment]
When I  had tried the solution you put in the doc, I always have the IDX10501 error.
I solved it using  this https://github.com/dotnet/AspNetCore.Docs/issues/16389#issuecomment-984468397

**Comment by Rick-Anderson** [comment]
@HaoK  should I replace @dandax sample with  https://github.com/dotnet/AspNetCore.Docs/issues/16389#issuecomment-984468397 ?

**Comment by HaoK** [comment]
I think so, we should definitely start encouraging more usage of policy schemes as this is exactly what it was designed for, to make it easier to compose and forward with extra logic like this.

**Comment by Rick-Anderson** [comment]
@dandax @egbakou @Sayan751 @jamir-araujo @automachina please review. Can some of the snippets be simplified? I'd appreciate a look at the entire topic and compare it to the [current published version](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-6.0)

**Comment by Rick-Anderson** [comment]
@HaoK  can you take a look? It's mostly updating the code to .NET 6 and moving the  embedded snippets in the MD file to a project file, + one new sample.

**Review by HaoK — APPROVED** [review]
Looks good overall, do we want to add a link or mention something about policy schemes, as I'm not sure how familiar most people would be with policy schemes and what options.ForwardDefaultSelector  does

File: aspnetcore/security/authorization/limitingidentitybyscheme.md [hunk 1]
```diff
@@ -41,102 +27,51 @@ In the preceding code, two authentication handlers have been added: one for cook
...
+
+In the preceding code, <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardDefaultSelector> is used to select a default scheme for the current request that authentication handlers should forward all authentication operations to by default. The default forwarding logic checks the most specific <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardAuthenticate>, <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardChallenge>, <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardForbid>, <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardSignIn>, and <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardSignIn> setting first, followed by checking the <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardDefaultSelector>, followed by <xref:Microsoft.AspNetCore.Authentication.AuthenticationSchemeOptions.ForwardDefault>. The first non null result is used as the target scheme to forward to. For more information, see <xref:security/authentication/policyschemes>.
```
Comment (Rick-Anderson) [inline aspnetcore/security/authorization/limitingidentitybyscheme.md:73]: @HaoK  See [AuthenticationSchemeOptions.ForwardDefaultSelector](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticationschemeoptions.forwarddefaultselector?view=aspnetcore-6.0) 
It lists Challenge/Forbid/SignIn/SignOut setting first
but it should be `ForwardChallenge`, `ForwardForbid`, `ForwardSignIn`, and `ForwardSignOut` setting first
They dropped the ***FORWARD*** prefix, right?

Comment (Rick-Anderson) [reply aspnetcore/security/authorization/limitingidentitybyscheme.md:73]: Added the links you requested.

**Review by Rick-Anderson — COMMENTED** [review]
@HaoK  quick reality question

File: aspnetcore/security/authorization/limitingidentitybyscheme.md (hunk 1 above)
Comment (HaoK) [reply aspnetcore/security/authorization/limitingidentitybyscheme.md]: Yep correct, Forward is a prefix for all of those settings

**Review by HaoK — APPROVED** [review]

**Pull request merged**